presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusSplitManager.java

```diff
@@ -0,0 +1,276 @@
+import com.facebook.presto.spi.TableNotFoundException;
+import com.facebook.presto.spi.connector.ConnectorSplitManager;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import io.airlift.units.Duration;
+
+import javax.inject.Inject;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.common.type.DateTimeEncoding.unpackMillisUtc;
+import static com.facebook.presto.common.type.TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
+import static com.facebook.presto.plugin.prometheus.PrometheusErrorCode.PROMETHEUS_UNKNOWN_ERROR;
+import static java.time.Instant.ofEpochMilli;
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusSplitManager
+        implements ConnectorSplitManager
+{
+    static final long OFFSET_MILLIS = 1L;
+    private final PrometheusClient prometheusClient;
+    private final PrometheusClock prometheusClock;
+
+    private final URI prometheusURI;
+    private final Duration maxQueryRangeDuration;
+    private final Duration queryChunkSizeDuration;
+
+    @Inject
+    public PrometheusSplitManager(PrometheusClient prometheusClient, PrometheusClock prometheusClock, PrometheusConnectorConfig config)
+    {
+        this.prometheusClient = requireNonNull(prometheusClient, "client is null");
+        this.prometheusClock = requireNonNull(prometheusClock, "prometheusClock is null");
+
+        requireNonNull(config, "config is null");
+        this.prometheusURI = config.getPrometheusURI();
+        this.maxQueryRangeDuration = config.getMaxQueryRangeDuration();
+        this.queryChunkSizeDuration = config.getQueryChunkSizeDuration();
+    }
+
+    /**
+     * Utility method to get the end times in decimal seconds that divide up the query into chunks
+     * The times will be used in queries to Prometheus like: `http://localhost:9090/api/v1/query?query=up[21d]&time=1568229904.000"`
+     * ** NOTE: Prometheus instant query wants the duration and end time specified.
+     * We use now() for the defaultUpperBound when none is specified, for instance, from predicate push down
+     *
+     * @param defaultUpperBound a LocalDateTime likely from PrometheusTimeMachine class for testability
+     * @return list of end times as decimal epoch seconds, like ["1568053244.143", "1568926595.321"]
+     */
+    protected static List<String> generateTimesForSplits(Instant defaultUpperBound, Duration maxQueryRangeDurationRequested, Duration queryChunkSizeDurationRequested,
+            PrometheusTableHandle tableHandle)
+    {
+        Optional<PrometheusPredicateTimeInfo> predicateRange = tableHandle.getPredicate()
```

Comment:
Shall we add precheck for `tableHandle.getPredicate()`? What if it's a fixed set like (1,2,3) ?

Reply:
good point, the predicate validation is done inside PrometheusSplitManager::determinePredicateTimes on filter to apply of `PrometheusColumnHandle.class`